### ISSUE TITLE
TFLite converter does not convert dilated conv to single conv op when spatial dimension is dynamic

### ISSUE THREAD

**Issue by WindQAQ** [tensorflow/tensorflow#46822]
### 1. System information

- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Colab
- TensorFlow installation (pip package or built from source): pip
- TensorFlow library (version, if pip package or github SHA, if built from source): 2.4 and nightly

### 2. Code

https://colab.research.google.com/drive/1U_rbD_tlvRmHjHXXFGMch_s_-2vl_cSf?usp=sharing

and visualize models with netron.

### 3. Failure after conversion

- Model converts successfully but the generated model does not have single dilated convolution op.

Although

https://github.com/tensorflow/tensorflow/blob/6b23dbc15988dd7bbfbdca7581f1cb2b85246e8c/tensorflow/compiler/mlir/lite/transforms/dilated_conv.h#L98-L102

says allow dynamic width and height, paddings in SpaceToBatchNDOp and crops in BatchToSpaceNDOp are not constant when input has dynamic spatial dim. Therefore,

https://github.com/tensorflow/tensorflow/blob/6b23dbc15988dd7bbfbdca7581f1cb2b85246e8c/tensorflow/compiler/mlir/lite/transforms/dilated_conv.h#L239-L253

fails to match this kind of pattern.

**Comment by abattery** [comment]
For dynamic dimensions, we can't make sure that the given inputs always are supported with the dilated conversion lowering or not. We have seen some cases that are not correctly performed with the certain inputs when there are dynamic dimensions. Once we can have a clear boundary between the supported and unsupported ranges even though the given inputs have partial dynamic dimensions, it will be possible to apply more cases. However, MLIR convert does have such rules yet.

**Comment by WindQAQ** [comment]
Thanks for clarification!

Our use case: In time sequence modeling, the input has dynamic spatial dim (batch, sequence, feature), we do multiple dilated conv to capture larger receptive field (finally followed by fixed size resizing). We've observed it's bottlenecked by conv2d on the hardware so it would be great to have support on dynamic dilated conv patterns. We actually have a workaround to move fixed sized resizing in front of dilated conv with a small accuracy loss but better speed.

My thought is to follow the codepath that keras does and bring the whole pattern into dag rewrite. Probably lengthy but doable I think. Is it possible to share the inputs that doesn't work correctly?

**Comment by abattery** [comment]
Here are some test cases related to that.

https://github.com/tensorflow/tensorflow/blob/a30d20b632b4ffbfd437ccf8ee205fef0917a3eb/tensorflow/compiler/mlir/lite/tests/dilated-conv.mlir#L329

**Comment by WindQAQ** [comment]
I am sorry to ask why this kind of inputs does not work correctly. It should work properly as least for dynamic batch dim.

**Comment by Saduf2019** [comment]
I am able to replicate the issue reported in tf 2.4 and nightly, please find the [gist here](https://colab.research.google.com/gist/Saduf2019/cf853efe2f0c63e3a176fe3bc8547e88/untitled528.ipynb).

**Comment by sushreebarsa** [comment]
Was able to reproduce the issue in TF v2.5,please find the gist [here](https://colab.research.google.com/gist/sushreebarsa/98b22f0c014d5446a7e561453dc5ba72/untitled48.ipynb)..Thanks !

**Comment by tilakrayal** [comment]
Hi,

Thank you for opening this issue. Since this issue has been open for a long time, the code/debug information for this issue may not be relevant with the current state of the code base.

The Tensorflow team is constantly improving the framework by fixing bugs and adding new features. We suggest you try the latest TensorFlow [version](https://pypi.org/project/tensorflow/) with the latest compatible hardware configuration which could potentially resolve the issue. If you are still facing the issue, please create a new GitHub issue with your latest findings, with all the debugging information which could help us investigate.

Please follow the [release notes](https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md) to stay up to date with the latest developments which are happening in the Tensorflow space.
